### PR TITLE
JIT: Fix STORE_DYN_BLK morphing

### LINKED ISSUE #73821: System.AccessViolationException in JIT compilation (.NET 7)
### Description

I am getting a `System.AccessViolationException` error when JIT compiling a method in .NET 7 preview 7.

### Reproduction Steps

Code to reproduce.

Program.cs:
```c#
using System;
using System.Runtime.CompilerServices;

namespace TestBug
{
    unsafe class Program
    {
        public struct Struct
        {
            public ushort f1;
            public byte f2;
            public byte f3;
        }

        public static Struct TestMethod()
        {
            Struct s;
            uint val = 0;
            ulong size = sizeof(uint);

            Unsafe.CopyBlockUnaligned(&s, &val, (uint) size);
            return s;
        }

        static void Main(string[] args)
        {
            TestMethod();
            Console.WriteLine("Ok");
        }
    }
}
```

TestBug.csproj:
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFrameworks>net6.0;net7.0</TargetFrameworks>
    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
    <LangVersion>9.0</LangVersion>
    <TieredCompilation>false</TieredCompilation>
  </PropertyGroup>

</Project>
```

### Expected behavior

```
Ok
```

### Actual behavior

```Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at TestBug.Program.TestMethod()
   at TestBug.Program.Main(System.String[])
```

### Regression?

Works fine in .NET 6

### Known Workarounds

_No response_

### Configuration

 OS Name:     Windows
 OS Version:  10.0.19044
 OS Platform: Windows

Host:
  Version:      7.0.0-preview.7.22375.6
  Architecture: x64
  Commit:       eecb028078

### Other information

_No response_

### PR DESCRIPTION
**Pull request by jakobbotsch**
Make sure to properly call `SetIndirExceptionFlags` and to properly mark defs (that will otherwise hit asserts in rationalization).

Fix #73821

### CODE REVIEW

**Comment by jakobbotsch** [comment]
Hmm nope, seems like the call is necessary. I will add it to `gtNewStructVal` instead.

**Review by SingleAccretion — APPROVED** [review]
Thank you for fixing this!

**Comment by jakobbotsch** [comment]
I've also noticed we don't set `GTF_VAR_DEF` on the local when we transform it to an ordinary assignment.
e.g.
```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
public static uint TestMethod()
{
    uint val1 = 0;
    uint val2 = 0;
    uint size = 2;
    Unsafe.CopyBlockUnaligned(&val2, &val1, size);
    return val2;
}
```

```scala
Assert failure(PID 49172 [0x0000c014], Thread: 23764 [0x5cd4]): Assertion failed 'tree->gtGetOp1()->gtFlags & GTF_VAR_DEF' in 'Runtime_73821:TestMethod():int' during 'Rationalize IR' (IL size 20; hash 0x11a894a2; FullOpts)

    File: C:\dev\dotnet\runtime2\src\coreclr\jit\rationalize.cpp Line: 333
    Image: C:\dev\dotnet\runtime2\artifacts\tests\coreclr\windows.x64.Checked\Tests\Core_Root\corerun.exe
```
Seems to me we should either have
```diff
@@ -8956,6 +8956,7 @@ GenTree* Compiler::fgMorphOneAsgBlockOp(GenTree* tree)
                 dest    = destLclVarTree;
                 asgType = destVarDsc->lvType;

+                destLclVarTree->gtFlags |= GTF_VAR_DEF;
                 // If the block operation had been a write to a local var of a small int type,
                 // of the exact size of the small int type, and the var is NormalizeOnStore,
                 // we would have labeled it GTF_VAR_USEASG, because the block operation wouldn't
```
or
```diff
@@ -1556,6 +1556,12 @@ GenTree* Compiler::fgMorphStoreDynBlock(GenTreeStoreDynBlk* tree)

             JITDUMP("MorphStoreDynBlock: transformed STORE_DYN_BLK into ASG(BLK, Data())\n");

+            GenTree* lclVarTree = fgIsIndirOfAddrOfLocal(lhs);
+            if (lclVarTree != nullptr)
+            {
+                lclVarTree->gtFlags |= GTF_VAR_DEF;
+            }
+
             return tree->OperIsCopyBlkOp() ? fgMorphCopyBlock(asg) : fgMorphInitBlock(asg);
         }
```

or both. Thoughts?

**Comment by SingleAccretion** [comment]
`GTF_VAR_DEF` does not need to be set for address-exposed locals (this is why `STORE_DYN_BLK` itself doesn't need to set it), so the assert in rationalization should be benign correctness-wise.

Still, it is probably best to still set it when transforming the `STORE_DYN_BLK`. There an argument to be made `gtNewAssignNode` should do it "all the way" with `fgAssignSetVarDef` instead of the current half-handling. Not sure how expensive that'd be.

**Comment by jakobbotsch** [comment]
It's a little odd to me that `fgMorphOneAsgBlockOp` sets `GTF_VAR_DEF` for the partial case but not for the whole def case though. It seems either both should be necessary or neither should be necessary?

**Comment by SingleAccretion** [comment]
> It's a little odd to me that `fgMorphOneAsgBlockOp` sets `GTF_VAR_DEF` for the partial case but not for the whole def case though. It seems either both should be necessary or neither should be necessary?

Hmm, I am not seeing the place where it only sets `VAR_DEF` for partial defs. It sets them both here:

https://github.com/dotnet/runtime/blob/55f5c7c0e0eb1d2a610b600da93c6415d7d46c98/src/coreclr/jit/morph.cpp#L8975-L8980

**Comment by jakobbotsch** [comment]
> Hmm, I am not seeing the place where it only sets `VAR_DEF` for partial defs. It sets them both here:

Right, what I mean is that it sets `VAR_DEF` there but not in the case above it (for whole defs). Do we expect `GTF_VAR_DEF` to not be set for partial defs here, but to be set for full defs? Should we have assert about the latter in the case above if so?

**Comment by SingleAccretion** [comment]
> Do we expect `GTF_VAR_DEF` to not be set for partial defs here, but to be set for full defs? Should we have assert about the latter in the case above if so?

The general expectation is that `VAR_DEF` will be set for both kinds of defs, because the function only looks at tree shapes that `DefinesLocal` recognizes and runs in post-order. So, yes, the fact code sets the flags for partial defs is a quirk and ought to be unnecessary.

**Comment by jakobbotsch** [comment]
Ok, good to know. I will add it to my todo list to add some asserts there then.

**Comment by jakobbotsch** [comment]
The timeouts seem unrelated.

cc @dotnet/jit-contrib PTAL @AndyAyersMS

File: src/coreclr/jit/gentree.cpp [hunk 1]
```diff
@@ -7541,6 +7541,8 @@ GenTree* Compiler::gtNewStructVal(ClassLayout* layout, GenTree* addr)
         blkNode = gtNewObjNode(layout, addr);
     }
 
+    blkNode->SetIndirExceptionFlags(this);
```
Comment (kunalspathak) [inline src/coreclr/jit/gentree.cpp:7544]: I am not too familiar with this, so could you elaborate how `SetIndirExceptionFlags` with the original problem?

Comment (jakobbotsch) [reply src/coreclr/jit/gentree.cpp:7544]: `SetIndirExceptionFlags` can only be called on trees that are indirs but this function can fold indirs of locals to `GT_LCL_VAR` nodes. So this change is just moving it into `gtNewStructVal` in the indir cases, similar to how it works for `gtNewIndir` today.

Comment (kunalspathak) [reply src/coreclr/jit/gentree.cpp:7544]: Got it.

**Review by kunalspathak — APPROVED** [review]
LGTM.

**Pull request merged**